{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af7bb5b-93f5-4bcf-860e-bc14f5ed1f06",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install --force-reinstall git+https://github.com/yhhc2/IsoRanker.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aaadd95-df6c-40c6-b1d5-ed9b292cb4fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "389cba39-78ff-459b-bb87-d9f10ea58d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import functions from the package\n",
    "from IsoRanker import (\n",
    "    filter_based_on_counts,\n",
    "    apply_hypothesis_test,\n",
    "    calculate_z_score,\n",
    "    NMD_test_statistic,\n",
    "    Noncyclo_Expression_Outlier_LOE,\n",
    "    Noncyclo_Expression_Outlier_GOE,\n",
    "    Cyclo_Expression_Outlier_GOE,\n",
    "    NMD_rare_steady_state_transcript,\n",
    "    Noncyclo_Allelic_Imbalance,\n",
    "    Cyclo_Allelic_Imbalance,\n",
    "    calculate_ranks_for_sample,\n",
    "    create_expression_matrix,\n",
    "    create_long_format,\n",
    "    process_hypothesis_test,\n",
    "    update_files_with_haplotype_info,\n",
    "    merge_tsvs_by_keyword, \n",
    "    process_vep_vcf, \n",
    "    merge_haplotype_data, \n",
    "    process_phenotype_data,\n",
    "    process_and_plot_pca,\n",
    "    analyze_isoforms,\n",
    "    process_pileup\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4f3eb3-4b64-42f8-9992-f7d35dec9fe1",
   "metadata": {},
   "source": [
    "## Assign input paths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e89e3928-4a63-4328-8aa6-eb9cea20a244",
   "metadata": {},
   "source": [
    "#### Option 1: If creating expression matrix from PacBio Isoseq pipeline, please uncomment and specify read_stat_path and sample_info_with_haplotype_location_path, otherwise comment out the whole block:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab8b7743-c370-48d3-bc90-cee55a8a419b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Option 1 ##\n",
    "\n",
    "# Provides every single read, which sample each read came from, and the isoform asssociated with each read\n",
    "read_stat_path = \"/mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker/examples/Expression_AllelicImbalance_NMD/Input/read_stats.txt\"\n",
    "\n",
    "# Assigns samples to patients, cyclo/non-cyclo, and haplotypes\n",
    "sample_info_with_haplotype_location_path = \"/mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker/examples/Expression_AllelicImbalance_NMD/Input/Sample_info.tsv\"\n",
    "\n",
    "# Update read_stat and sample_info file with haplotype information\n",
    "update_files_with_haplotype_info(sample_info_with_haplotype_location_path, read_stat_path, \".\")\n",
    "\n",
    "# Provides every single read, which sample each read came from, and the isoform asssociated with each read\n",
    "read_stat_path = \"updated_read_stats.txt.gz\" # This should NOT be edited if using option 1\n",
    "sample_info_path = \"updated_sample_info.tsv.gz\" # This should NOT be edited if using option 1\n",
    "\n",
    "# Assigns samples to patients, cyclo/non-cyclo, and haplotypes\n",
    "sample_info = pd.read_csv(sample_info_path, sep=\"\\t\")\n",
    "\n",
    "# Create the expression matrix and save it to a file\n",
    "expression_matrix = create_expression_matrix(read_stat_path, output_file=\"expression_matrix.tsv.gz\") # This should NOT be edited if using option 1\n",
    "\n",
    "## Option 1 ##"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf0089d-7d78-4a58-b847-0fa828cee8e0",
   "metadata": {},
   "source": [
    "#### Option 2: If using expression matrix directly, please uncomment and specify expression_matrix_path and sample_info_path, otherwise comment out the whole block.\n",
    "- Please see Output/intermediate/expression_matrix.tsv.gz for an example of how to format expression matrix.\n",
    "- Please see Output/intermediate/updated_sample_info.tsv.gz for an example of how to format sample info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ad329f-4a86-4631-91bd-47cec59efe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Option 2 ##\n",
    "\n",
    "# expression_matrix_path = \"/mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker/examples/Expression_AllelicImbalance_NMD/Output/intermediate/expression_matrix.tsv.gz\"\n",
    "# sample_info_path = \"/mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker/examples/Expression_AllelicImbalance_NMD/Output/intermediate/updated_sample_info.tsv.gz\"\n",
    "\n",
    "# expression_matrix = pd.read_csv(expression_matrix_path, compression = \"gzip\", index_col=0, sep=\"\\t\")\n",
    "# sample_info = pd.read_csv(sample_info_path, compression = \"gzip\", sep=\"\\t\")\n",
    "\n",
    "## Option 2 ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21232689-4512-49bd-99c6-dd73ed0e3640",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Provides pigeon annotation for isoforms\n",
    "classification_path = \"/mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker/examples/Expression_AllelicImbalance_NMD/Input/filtered_classification.txt\"\n",
    "classification_data = pd.read_csv(classification_path, sep=\"\\t\")\n",
    "\n",
    "# Provide omim information for genes\n",
    "genemap_path = \"/mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker_testing/genemap2.txt\"\n",
    "genemap = pd.read_csv(genemap_path, sep='\\t', skiprows=3) # Read the file, skipping the first 3 rows\n",
    "genemap = genemap[genemap['Approved Gene Symbol'].notnull()]\n",
    "\n",
    "# Reference file for SRSF6 QC pileup\n",
    "reference_fasta_path = \"/gscratch/stergachislab/assemblies/simple-names/hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb1ee996-266d-4e3f-bf73-15fb8918295b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8007cb48-f20b-439d-82c4-3bb8e3f6d138",
   "metadata": {},
   "source": [
    "## Convert isoform matrix to long format, then add gene information to enable gene-level analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dce79e2-f105-4dbd-882b-c85b83714177",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the long-format dataFrame and adding cyclo/noncyclo as well as haplotype labels and calculate TPM\n",
    "long_format_df = create_long_format(expression_matrix, sample_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e60edba-252d-49df-98a3-38a564315cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select only the 'isoform' and 'associated_gene' columns from classification_data\n",
    "classification_subset = classification_data[['isoform', 'associated_gene']]\n",
    "# Merge the classification subset with the long_format_df\n",
    "long_format_annotated = long_format_df.merge(\n",
    "    classification_subset,\n",
    "    left_on=\"Isoform\",   # Match isoform IDs in long_format_df\n",
    "    right_on=\"isoform\",  # Match isoform IDs in classification_subset\n",
    "    how=\"left\"           # Keep all rows from long_format_df, even if there's no match in classification_subset\n",
    ").drop(columns=[\"isoform\"])  # Drop redundant 'isoform' column from classification_subset\n",
    "\n",
    "long_format_annotated.to_csv(\"long_format_annotated.tsv.gz\", index=False, compression = \"gzip\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78e6fbe3-4082-4567-858b-9658caf15d3f",
   "metadata": {},
   "source": [
    "## Calculate test stat and rankings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d55d60b-0c43-4847-9cf5-88b1dde4dfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate for all hypothesis tests\n",
    "test_stat_funcs = [\n",
    "    (\"NMD\", NMD_test_statistic),\n",
    "    (\"Noncyclo_LOE\", Noncyclo_Expression_Outlier_LOE),\n",
    "    (\"Noncyclo_GOE\", Noncyclo_Expression_Outlier_GOE),\n",
    "    (\"Cyclo_GOE\", Cyclo_Expression_Outlier_GOE),\n",
    "    (\"NMD_rare_steady_state_transcript\", NMD_rare_steady_state_transcript),\n",
    "    (\"Nonyclo_Allelic_Imbalance\", Noncyclo_Allelic_Imbalance),\n",
    "    (\"Cyclo_Allelic_Imbalance\", Cyclo_Allelic_Imbalance)\n",
    "]\n",
    "\n",
    "# Store full results to generate lookup table\n",
    "full_ranked_gene_data = []\n",
    "\n",
    "#Gene level\n",
    "for test_name, test_func in test_stat_funcs:\n",
    "        print(f\"Processing test statistic: {test_name}\")\n",
    "\n",
    "        # Apply the process_hypothesis_test function\n",
    "        ranked_data = process_hypothesis_test(\n",
    "            filtered_data=long_format_annotated, \n",
    "            group_col='Isoform', \n",
    "            test_statistic_func=test_func, \n",
    "            gene_group_col='associated_gene', \n",
    "            gene_level=True, \n",
    "            bin_proportion=0.01, \n",
    "            filter_before_ranking=True, \n",
    "            filter_count_threshold=10)\n",
    "\n",
    "        # Append tuple (test_name, ranked_data) to the list\n",
    "        full_ranked_gene_data.append((test_name, ranked_data))\n",
    "\n",
    "        filtered_ranked_data = ranked_data[ranked_data[\"rank_top_99_5_percentile\"] <= 25]\n",
    "\n",
    "        # Add OMIM data to genes\n",
    "        filtered_ranked_data = filtered_ranked_data.merge(\n",
    "            genemap[['Approved Gene Symbol', 'Phenotypes']],  # Select relevant columns from genemap\n",
    "            how='left',  # Perform a left join to keep all rows from filtered_ranked_data\n",
    "            left_on='associated_gene',  # Column in filtered_ranked_data to join on\n",
    "            right_on='Approved Gene Symbol'  # Column in genemap to join on\n",
    "        )\n",
    "        # Drop the 'Approved Gene Name' column if it is no longer needed\n",
    "        filtered_ranked_data = filtered_ranked_data.drop(columns=['Approved Gene Symbol'])\n",
    "\n",
    "        # Save the results to a tsv file\n",
    "        output_dir = \"\"\n",
    "        output_file = os.path.join(output_dir, f\"{test_name}_gene_top_ranked_data.tsv.gz\")\n",
    "        filtered_ranked_data.to_csv(output_file, index=False, compression = \"gzip\", sep=\"\\t\")\n",
    "        print(f\"Results saved to {output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0e3d228-f632-47e6-857f-ef1601a01ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_stat_funcs = [\n",
    "    (\"NMD\", NMD_test_statistic),\n",
    "    (\"Noncyclo_LOE\", Noncyclo_Expression_Outlier_LOE),\n",
    "    (\"Noncyclo_GOE\", Noncyclo_Expression_Outlier_GOE),\n",
    "    (\"Cyclo_GOE\", Cyclo_Expression_Outlier_GOE)\n",
    "]\n",
    "\n",
    "#Isoform level\n",
    "for test_name, test_func in test_stat_funcs:\n",
    "        print(f\"Processing test statistic: {test_name}\")\n",
    "\n",
    "        # Apply the process_hypothesis_test function\n",
    "        ranked_data = process_hypothesis_test(\n",
    "            filtered_data=long_format_annotated, \n",
    "            group_col='Isoform', \n",
    "            test_statistic_func=test_func, \n",
    "            gene_group_col='associated_gene', \n",
    "            gene_level=False, \n",
    "            bin_proportion=0.01, \n",
    "            filter_before_ranking=True, \n",
    "            filter_count_threshold=10)\n",
    "\n",
    "        filtered_ranked_data = ranked_data[ranked_data[\"rank_top_99_5_percentile\"] <= 25]\n",
    "\n",
    "        # Add OMIM data to genes\n",
    "        filtered_ranked_data = filtered_ranked_data.merge(\n",
    "            genemap[['Approved Gene Symbol', 'Phenotypes']],  # Select relevant columns from genemap\n",
    "            how='left',  # Perform a left join to keep all rows from filtered_ranked_data\n",
    "            left_on='associated_gene',  # Column in filtered_ranked_data to join on\n",
    "            right_on='Approved Gene Symbol'  # Column in genemap to join on\n",
    "        )\n",
    "        # Drop the 'Approved Gene Name' column if it is no longer needed\n",
    "        filtered_ranked_data = filtered_ranked_data.drop(columns=['Approved Gene Symbol'])\n",
    "\n",
    "        # Save the results to a tsv file\n",
    "        output_dir = \"\"\n",
    "        output_file = os.path.join(output_dir, f\"{test_name}_isoform_top_ranked_data.tsv.gz\")\n",
    "        filtered_ranked_data.to_csv(output_file, index=False, compression = \"gzip\", sep=\"\\t\")\n",
    "        print(f\"Results saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2de41b6-1766-4ae3-b7d7-a98c6091afe9",
   "metadata": {},
   "source": [
    "## Combine all output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "486cdd7a-46f5-4a08-a1fb-99726ebf2f57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isoform\n",
    "directory = \".\"\n",
    "keyword = \"isoform\" \n",
    "output_tsv = os.path.join(directory, f\"merged_ranked_{keyword}.tsv.gz\")\n",
    "merge_tsvs_by_keyword(directory, keyword, output_tsv)\n",
    "\n",
    "# Gene\n",
    "directory = \".\"\n",
    "keyword = \"gene\"\n",
    "output_tsv = os.path.join(directory, f\"merged_ranked_{keyword}.tsv.gz\")\n",
    "merge_tsvs_by_keyword(directory, keyword, output_tsv)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82ca6a17-3a04-45a1-a09d-976e386ae418",
   "metadata": {},
   "source": [
    "## Lookup tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6700e753-9159-4ccc-b823-15ae7f77705c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collapse long_format_annotated by summing TPM values per Sample and Gene\n",
    "sample_gene_rankings_lookup_table = long_format_annotated.groupby([\"Sample\", \"associated_gene\"], as_index=False).agg(\n",
    "    {\"Cyclo_TPM\": \"sum\", \"Noncyclo_TPM\": \"sum\"}\n",
    ")\n",
    "\n",
    "# Now merge the ranked gene-level data\n",
    "for test_name, df in full_ranked_gene_data:\n",
    "    # Rename rank column to be test-specific\n",
    "    df_renamed = df.rename(columns={\"rank_top_99_5_percentile\": f\"{test_name}_rank_top_99_5_percentile\"})\n",
    "\n",
    "    # Keep only relevant columns\n",
    "    df_renamed = df_renamed[[\"Sample\", \"associated_gene\", f\"{test_name}_rank_top_99_5_percentile\"]]\n",
    "\n",
    "    # Merge into merged_df using outer join\n",
    "    sample_gene_rankings_lookup_table = pd.merge(sample_gene_rankings_lookup_table, df_renamed, on=[\"Sample\", \"associated_gene\"], how=\"outer\")\n",
    "\n",
    "# Save as a compressed tsv (gzip format)\n",
    "sample_gene_rankings_lookup_table.to_csv(\"sample_gene_rankings_lookup_table.tsv.gz\", index=False, compression=\"gzip\", sep=\"\\t\")\n",
    "\n",
    "# Group by gene (associated_gene) and compute median, Q1 (25th percentile), and Q3 (75th percentile)\n",
    "gene_coverage_lookup_table = sample_gene_rankings_lookup_table.groupby(\"associated_gene\").agg(\n",
    "    Cyclo_TPM_median=(\"Cyclo_TPM\", \"median\"),\n",
    "    Cyclo_TPM_Q1=(\"Cyclo_TPM\", lambda x: x.quantile(0.25)),  # 25th percentile\n",
    "    Cyclo_TPM_Q3=(\"Cyclo_TPM\", lambda x: x.quantile(0.75)),  # 75th percentile\n",
    "    Cyclo_TPM_min=(\"Cyclo_TPM\", \"min\"),  # Minimum value\n",
    "    Cyclo_TPM_max=(\"Cyclo_TPM\", \"max\"),  # Maximum value\n",
    "\n",
    "    Noncyclo_TPM_median=(\"Noncyclo_TPM\", \"median\"),\n",
    "    Noncyclo_TPM_Q1=(\"Noncyclo_TPM\", lambda x: x.quantile(0.25)),  # 25th percentile\n",
    "    Noncyclo_TPM_Q3=(\"Noncyclo_TPM\", lambda x: x.quantile(0.75)),  # 75th percentile\n",
    "    Noncyclo_TPM_min=(\"Noncyclo_TPM\", \"min\"),  # Minimum value\n",
    "    Noncyclo_TPM_max=(\"Noncyclo_TPM\", \"max\")   # Maximum value\n",
    ").reset_index()\n",
    "\n",
    "gene_coverage_lookup_table.to_csv(\"gene_coverage_lookup_table.tsv.gz\", index=False, compression=\"gzip\", sep=\"\\t\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d234832b-47b0-4bdc-962f-04c022296a01",
   "metadata": {},
   "source": [
    "## QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4918c31a-87a8-4e99-8e68-c92d22677fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "# PCA\n",
    "###################################\n",
    "\n",
    "pca_results = process_and_plot_pca(long_format_annotated, output_pdf=\"pca_plot.pdf\", grouping_col = \"associated_gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de83ab30-d708-4338-8c66-cd30b8311637",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "# Gene diversity\n",
    "###################################\n",
    "\n",
    "analyze_isoforms(long_format_annotated, \"gene_diversity.tsv.gz\", \"associated_gene\")\n",
    "\n",
    "###################################\n",
    "# Isoform diversity\n",
    "###################################\n",
    "\n",
    "analyze_isoforms(long_format_annotated, \"isoform_diversity.tsv.gz\", \"Isoform\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf603643-28fe-4b5a-a59b-e552a79c60ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "# Gene diversity plot\n",
    "###################################\n",
    "\n",
    "df = pd.read_csv(\"gene_diversity.tsv.gz\", compression = \"gzip\", sep=\"\\t\")\n",
    "\n",
    "# Drop 'Cyclo Total Reads' and 'Noncyclo Total Reads' columns\n",
    "df_filtered = df.drop(columns=[\"Cyclo Total Reads\", \"Noncyclo Total Reads\"])\n",
    "\n",
    "# Define colors for Cyclo and Noncyclo categories\n",
    "cyclo_color = \"red\"\n",
    "noncyclo_color = \"blue\"\n",
    "\n",
    "# Determine a common y-axis limit for all plots\n",
    "y_max = df_filtered.drop(columns=[\"Sample\"]).max().max()\n",
    "\n",
    "# Create subplots for each sample\n",
    "num_samples = len(df_filtered[\"Sample\"])\n",
    "fig, axes = plt.subplots(nrows=num_samples, figsize=(12, num_samples * 3), sharex=True, sharey=True)\n",
    "\n",
    "# If only one sample, make axes iterable\n",
    "if num_samples == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "# Loop through each sample and create a separate bar plot\n",
    "for ax, sample in zip(axes, df_filtered[\"Sample\"]):\n",
    "    sample_data = df_filtered[df_filtered[\"Sample\"] == sample].drop(columns=[\"Sample\"]).T\n",
    "    colors = [cyclo_color if \"Cyclo\" in col else noncyclo_color for col in sample_data.index]\n",
    "    \n",
    "    ax.bar(sample_data.index, sample_data.iloc[:, 0], color=colors)\n",
    "    ax.set_title(f\"Sample: {sample}\")\n",
    "    ax.set_ylabel(\"Count\")\n",
    "    ax.set_ylim(0, y_max)  # Set common y-axis limit\n",
    "    ax.tick_params(axis=\"x\", rotation=90)\n",
    "\n",
    "# Formatting\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save to PDF\n",
    "plt.savefig(\"gene_diversity.pdf\", format=\"pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c0c9d7f-96f4-46eb-8c86-f2e84faef79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "# Isoform diversity plot\n",
    "###################################\n",
    "\n",
    "df = pd.read_csv(\"isoform_diversity.tsv.gz\", compression = \"gzip\", sep=\"\\t\")\n",
    "\n",
    "# Drop 'Cyclo Total Reads' and 'Noncyclo Total Reads' columns\n",
    "df_filtered = df.drop(columns=[\"Cyclo Total Reads\", \"Noncyclo Total Reads\"])\n",
    "\n",
    "# Define colors for Cyclo and Noncyclo categories\n",
    "cyclo_color = \"red\"\n",
    "noncyclo_color = \"blue\"\n",
    "\n",
    "# Determine a common y-axis limit for all plots\n",
    "y_max = df_filtered.drop(columns=[\"Sample\"]).max().max()\n",
    "\n",
    "# Create subplots for each sample\n",
    "num_samples = len(df_filtered[\"Sample\"])\n",
    "fig, axes = plt.subplots(nrows=num_samples, figsize=(12, num_samples * 3), sharex=True, sharey=True)\n",
    "\n",
    "# If only one sample, make axes iterable\n",
    "if num_samples == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "# Loop through each sample and create a separate bar plot\n",
    "for ax, sample in zip(axes, df_filtered[\"Sample\"]):\n",
    "    sample_data = df_filtered[df_filtered[\"Sample\"] == sample].drop(columns=[\"Sample\"]).T\n",
    "    colors = [cyclo_color if \"Cyclo\" in col else noncyclo_color for col in sample_data.index]\n",
    "    \n",
    "    ax.bar(sample_data.index, sample_data.iloc[:, 0], color=colors)\n",
    "    ax.set_title(f\"Sample: {sample}\")\n",
    "    ax.set_ylabel(\"Count\")\n",
    "    ax.set_ylim(0, y_max)  # Set common y-axis limit\n",
    "    ax.tick_params(axis=\"x\", rotation=90)\n",
    "\n",
    "# Formatting\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save to PDF\n",
    "plt.savefig(\"isoform_diversity.pdf\", format=\"pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c2bee67-c652-4449-8c8f-6de862934518",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"bam_file\" in sample_info.columns and sample_info[\"bam_file\"].notna().any():\n",
    "    \n",
    "    ###################################\n",
    "    # SRSF6\n",
    "    ###################################\n",
    "    process_pileup(df=sample_info, reference_fasta=reference_fasta_path, chromosome=\"chr20\", position=43459200, output_file=\"SRSF6.tsv.gz\")\n",
    "\n",
    "    ###################################\n",
    "    # SRSF6 plot \n",
    "    ###################################\n",
    "    SRSF6_df = pd.read_csv(\"SRSF6.tsv.gz\", compression=\"gzip\", sep=\"\\t\")\n",
    "\n",
    "    # Sort DataFrame by source\n",
    "    df_sorted = SRSF6_df.sort_values(by=\"Source\")\n",
    "\n",
    "    # Define colors based on condition\n",
    "    colors = [\"red\" if \"_cyclo\" in source else \"blue\" for source in df_sorted[\"Source\"]]\n",
    "    \n",
    "    # Plot bar chart\n",
    "    plt.figure(figsize=(15, 9))\n",
    "    sns.barplot(data=df_sorted, x=\"Source\", y=\"Exonic_Proportion\", edgecolor=\"black\", palette=colors)\n",
    "    \n",
    "    # Labeling and formatting\n",
    "    plt.xlabel(\"Source\")\n",
    "    plt.ylabel(\"Exonic Proportion\")\n",
    "    plt.title(\"Bar Plot of Exonic Proportion by Source\")\n",
    "    plt.xticks(rotation=45, fontsize=7, ha=\"right\")  # Adjust alignment to prevent cutoff\n",
    "\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    # Save to PDF\n",
    "    plt.savefig(\"SRSF6_exonic_proportion.pdf\", format=\"pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ac03b99-a29f-4f37-a9d8-8afd837e901c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute deltas between every pair of rows (0-1, 2-3, etc.)\n",
    "deltas = []\n",
    "for i in range(0, len(df_sorted), 2):\n",
    "    source = df_sorted.iloc[i][\"Source\"] + \"_\" + df_sorted.iloc[i + 1][\"Source\"]\n",
    "    delta = (df_sorted.iloc[i][\"Exonic_Proportion\"] - df_sorted.iloc[i + 1][\"Exonic_Proportion\"])/df_sorted.iloc[i][\"Exonic_Proportion\"] \n",
    "    deltas.append({\"Source\": source, \"Exonic_Delta\": delta})\n",
    "\n",
    "# Create new DataFrame\n",
    "delta_df = pd.DataFrame(deltas)\n",
    "\n",
    "# Optional: Set a style\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "# Create the boxplot\n",
    "plt.figure(figsize=(8, 5))\n",
    "ax = sns.boxplot(data=delta_df, y=\"Exonic_Delta\", color=\"lightgray\", width=0.3)\n",
    "\n",
    "# Overlay the swarmplot\n",
    "sns.swarmplot(data=delta_df, y=\"Exonic_Delta\", color=\"black\", size=8)\n",
    "\n",
    "# Set y-axis limits\n",
    "ax.set_ylim(-1, 1.1)\n",
    "\n",
    "# Optional: Add x-axis label\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"Normalized Exonic Delta\")\n",
    "ax.set_title(\"Normalized Exonic Delta per Patient\")\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save to PDF\n",
    "plt.savefig(\"SRSF6_exonic_proportion_normalized_delta.pdf\", format=\"pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20fa931a-762b-46e0-8fd3-c000536195e7",
   "metadata": {},
   "source": [
    "## Organizing output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b96031cf-574a-4e08-97df-36299bf1907b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "\n",
    "# Define folder names\n",
    "OUTPUT_FOLDER = \"Output\"\n",
    "QC_FOLDER = os.path.join(OUTPUT_FOLDER, \"qc\")\n",
    "BROWSER_FOLDER = os.path.join(OUTPUT_FOLDER, \"browser\")\n",
    "LOOKUP_TABLES_FOLDER = os.path.join(BROWSER_FOLDER, \"lookup_tables\")\n",
    "COMBINED_RESULTS_FOLDER = os.path.join(BROWSER_FOLDER, \"combined_results\")\n",
    "SEPARATED_RESULTS_FOLDER = os.path.join(BROWSER_FOLDER, \"separated_results\")\n",
    "INTERMEDIATE_FOLDER = os.path.join(OUTPUT_FOLDER, \"intermediate\")\n",
    "\n",
    "# Ensure directories exist\n",
    "os.makedirs(OUTPUT_FOLDER, exist_ok=True)\n",
    "os.makedirs(QC_FOLDER, exist_ok=True)\n",
    "os.makedirs(BROWSER_FOLDER, exist_ok=True)\n",
    "os.makedirs(LOOKUP_TABLES_FOLDER, exist_ok=True)\n",
    "os.makedirs(COMBINED_RESULTS_FOLDER, exist_ok=True)\n",
    "os.makedirs(SEPARATED_RESULTS_FOLDER, exist_ok=True)\n",
    "os.makedirs(INTERMEDIATE_FOLDER, exist_ok=True)\n",
    "\n",
    "\n",
    "# Define file categories\n",
    "qc_files = {\n",
    "    \"pca_plot.pdf\",\n",
    "    \"gene_diversity.tsv.gz\",\n",
    "    \"isoform_diversity.tsv.gz\",\n",
    "    \"SRSF6.tsv.gz\",\n",
    "    \"SRSF6_exonic_proportion.pdf\",\n",
    "    \"SRSF6_exonic_proportion_normalized_delta.pdf\",\n",
    "    \"isoform_diversity.pdf\",\n",
    "    \"gene_diversity.pdf\"\n",
    "}\n",
    "\n",
    "lookup_table_files = {\n",
    "    \"sample_gene_rankings_lookup_table.tsv.gz\",\n",
    "    \"gene_coverage_lookup_table.tsv.gz\"\n",
    "}\n",
    "\n",
    "combined_results_files = {\n",
    "    \"merged_ranked_gene_with_phenotype.tsv.gz\",\n",
    "    \"merged_ranked_isoform_with_phenotype.tsv.gz\",\n",
    "    \"merged_ranked_gene.tsv.gz\",\n",
    "    \"merged_ranked_isoform.tsv.gz\"\n",
    "}\n",
    "\n",
    "separated_results_files = {\n",
    "    \"Cyclo_Allelic_Imbalance_gene_top_ranked_data.tsv.gz\",\n",
    "    \"Cyclo_GOE_gene_top_ranked_data.tsv.gz\",\n",
    "    \"Cyclo_GOE_isoform_top_ranked_data.tsv.gz\",\n",
    "    \"NMD_gene_top_ranked_data.tsv.gz\",\n",
    "    \"NMD_isoform_top_ranked_data.tsv.gz\",\n",
    "    \"NMD_rare_steady_state_transcript_gene_top_ranked_data.tsv.gz\",\n",
    "    \"Noncyclo_GOE_gene_top_ranked_data.tsv.gz\",\n",
    "    \"Noncyclo_GOE_isoform_top_ranked_data.tsv.gz\",\n",
    "    \"Noncyclo_LOE_gene_top_ranked_data.tsv.gz\",\n",
    "    \"Noncyclo_LOE_isoform_top_ranked_data.tsv.gz\",\n",
    "    \"Nonyclo_Allelic_Imbalance_gene_top_ranked_data.tsv.gz\"\n",
    "}\n",
    "\n",
    "browser_files = lookup_table_files | combined_results_files | separated_results_files\n",
    "\n",
    "# Get all `.gz` files in the current directory\n",
    "all_gz_files = {f for f in os.listdir() if f.endswith(\".gz\")}\n",
    "\n",
    "# Find files that should go into intermediate (everything not in QC or Browser)\n",
    "intermediate_files = all_gz_files - qc_files - browser_files\n",
    "\n",
    "# Function to move files\n",
    "def move_files(file_list, destination_folder):\n",
    "    for file in file_list:\n",
    "        if os.path.exists(file):  # Ensure the file exists before moving\n",
    "            shutil.move(file, os.path.join(destination_folder, file))\n",
    "        else:\n",
    "            print(f\"Warning: {file} not found, skipping.\")\n",
    "\n",
    "# Move files to their respective folders\n",
    "move_files(qc_files, QC_FOLDER)\n",
    "move_files(lookup_table_files, LOOKUP_TABLES_FOLDER)\n",
    "move_files(combined_results_files, COMBINED_RESULTS_FOLDER)\n",
    "move_files(separated_results_files, SEPARATED_RESULTS_FOLDER)\n",
    "move_files(intermediate_files, INTERMEDIATE_FOLDER)\n",
    "\n",
    "print(\"File organization complete!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49ac75bd-7e8c-4943-bf0b-a0c8d74d9f69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3642bdbe-d4ae-4f0e-8163-8437d46d7b73",
   "metadata": {},
   "source": [
    "## Zipping/Unzipping all the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf81f4b4-aaa2-4a1a-b87d-8e6bb53ca4b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "def unzip_gz_files_recursively(directory, output_dir):\n",
    "    \"\"\"\n",
    "    Recursively unzips all .gz files from `directory` and its subdirectories, \n",
    "    preserving the folder structure in `output_dir`.\n",
    "    \n",
    "    Parameters:\n",
    "    - directory (str): Root directory to search for .gz files.\n",
    "    - output_dir (str): Destination directory where extracted files will be saved.\n",
    "    \"\"\"\n",
    "    for root, _, files in os.walk(directory):  # Recursively walk through directories\n",
    "        for file in files:\n",
    "            if file.endswith(\".gz\"):\n",
    "                gz_path = os.path.join(root, file)\n",
    "                \n",
    "                # Preserve subdirectory structure in the output directory\n",
    "                relative_path = os.path.relpath(root, directory)  # Get relative path\n",
    "                target_dir = os.path.join(output_dir, relative_path)\n",
    "                os.makedirs(target_dir, exist_ok=True)  # Create subdirectories if needed\n",
    "                \n",
    "                output_path = os.path.join(target_dir, file[:-3])  # Remove .gz extension\n",
    "                \n",
    "                # Extract file\n",
    "                with gzip.open(gz_path, 'rb') as gz_file, open(output_path, 'wb') as out_file:\n",
    "                    shutil.copyfileobj(gz_file, out_file)\n",
    "\n",
    "                print(f\"Extracted: {gz_path} -> {output_path}\")\n",
    "\n",
    "# Example usage\n",
    "unzip_gz_files_recursively(\n",
    "    \".\", \n",
    "    \"Output_Unzipped\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03fb8ab9-316b-482c-a51e-7844381123ef",
   "metadata": {},
   "source": [
    "## Unit tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3784dc31-64e2-47b3-8102-8d0537efa0d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run unit tests\n",
    "# !python -m unittest discover -s /mmfs1/gscratch/stergachislab/yhhc/projects/IsoRanker/tests -p \"test_all.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3beba21-1ad6-48b7-b7e6-a16915b88930",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e1f4638-4253-4557-ad09-f3a3a65ce096",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63083257-2648-4164-a4f7-d36a1aa8f7dd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
